cpp/src/arrow/compute/exec/spilling_join.h

```diff
@@ -0,0 +1,100 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <bitset>
+
+#include "arrow/compute/exec/query_context.h"
+#include "arrow/compute/exec/hash_join.h"
+#include "arrow/compute/exec/accumulation_queue.h"
+
+namespace arrow
+{
+    namespace compute
+    {
+        class SpillingHashJoin
+        {
+        public:
+            using OutputBatchCallback = std::function<void(int64_t, ExecBatch)>;
+            using BuildFinishedCallback = std::function<Status(size_t)>;
+            using FinishedCallback = std::function<void(int64_t)>;
+            using RegisterTaskGroupCallback = std::function<int(
+                std::function<Status(size_t, int64_t)>, std::function<Status(size_t)>)>;
+            using StartTaskGroupCallback = std::function<Status(int, int64_t)>;
+            using PauseProbeSideCallback = std::function<void(int)>;
+            using ResumeProbeSideCallback = std::function<void(int)>;
+            using AbortContinuationImpl = std::function<void()>;
+
+            struct CallbackRecord
+            {
+                OutputBatchCallback output_batch_callback;
+                BuildFinishedCallback build_finished_callback;
+                FinishedCallback finished_callback;
+                RegisterTaskGroupCallback register_task_group_;
+                StartTaskGroupCallback start_task_group_callback;
+                PauseProbeSideCallback pause_probe_side_callback;
+                AbortContinuationImpl abort_callback;
+            };
```

Comment:
Can we just use a pure virtual class at this point?

```
class HashJoinExternals {
  virtual void OutputBatch(int64_t, ExecBatch) = 0;
  // ...
};
```

Reply:
I personally find these pure-virtual classes cumbersome to deal with as they remove the callback record from being near the site where I invoke `Init`. They're also less flexible and don't let me reuse functions (like HashJoinImpl and SpillingHashJoin reuse a lot of the same callbacks, I can just assign the same stuff between the two callback records).